geoh5py/shared/validators.py

```diff
@@ -0,0 +1,241 @@
+#  Copyright (c) 2022 Mira Geoscience Ltd.
+#
+#  This file is part of geoh5py.
+#
+#  geoh5py is free software: you can redistribute it and/or modify
+#  it under the terms of the GNU Lesser General Public License as published by
+#  the Free Software Foundation, either version 3 of the License, or
+#  (at your option) any later version.
+#
+#  geoh5py is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU Lesser General Public License for more details.
+#
+#  You should have received a copy of the GNU Lesser General Public License
+#  along with geoh5py.  If not, see <https://www.gnu.org/licenses/>.
+#
+#  This file is part of geoapps.
+#
+#  geoapps is distributed under the terms and conditions of the MIT License
+#  (see LICENSE file at the root of this source code package).
+
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import Any
+from uuid import UUID
+
+import numpy as np
+
+from geoh5py.groups import PropertyGroup
+from geoh5py.shared import Entity
+from geoh5py.shared.exceptions import (
+    AssociationValidationError,
+    PropertyGroupValidationError,
+    RequiredValidationError,
+    ShapeValidationError,
+    TypeValidationError,
+    UUIDStringValidationError,
+    UUIDValidationError,
+    ValueValidationError,
+)
+from geoh5py.shared.utils import iterable
+from geoh5py.workspace import Workspace
+
+
+class BaseValidator(ABC):
+    """
+    Base class for validators
+    """
+
+    def __init__(self, **kwargs):
+
+        self._validation_type = ""
+
+        for key, value in kwargs.items():
+            if hasattr(self, key):
+                setattr(self, key, value)
+
+    @classmethod
+    @abstractmethod
+    def validate(cls, name: str, value: Any, valid: Any):
+        """
+        Custom validation function.
+        """
+        raise NotImplementedError(
+            "The 'validate' method must be implemented by the sub-class. "
+            f"Must contain a 'name' {name}, 'value' {value} and 'valid' {valid} argument."
+        )
+
+    @property
+    @abstractmethod
+    def validation_type(self):
+        ...
+
+    def __call__(self, *args):
+        if hasattr(self, "validate"):
```

Comment:
It's more a protection against kwargs that aren't defined on the class

Reply:
Not sure I know what you mean.  Since the attribute (method) 'validate' is decorated with an abstractmethod in the abstract base class, there's no such thing as a validator that inherits this `__call__` method that doesn't also have a valdiate method

Reply:
Oh yep sorry, was reading it from the __init__. True shouldn't need it.